### PR TITLE
ENH: stats.bootstrap: add BCa method for multi-sample statistic

### LINKED ISSUE #16433: ENH: scipy.stats.bootstrap() should do BCa for multivariate statistics too
### Is your feature request related to a problem? Please describe.

I've seen this several times while looking at datasets. In R I can do BCa correction any time. When I try to do the same analysis with Python and I want to apply the BCa correction, it only works for univariate.

### Describe the solution you'd like.

Would be really nice to have BCa available for all analyses.

### Describe alternatives you've considered.

R-lang - I mean, it works already.

### Additional context (e.g. screenshots, GIFs)

_No response_

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-16433
Addresses item in gh-13371

#### What does this implement/fix?
Currently, `stats.bootstrap` only supports `method='BCa'` for single-sample statistics. This adds support for multi-sample statistics.

I need to add a test. @FlorinAndrei can you help me devise one? One way to do this would be to provide output from R's `boot.ci`  to compare against. Ideally, this would be performed with many (say ~100k resamples) to avoid discrepancies due to the choice of random samples, and it would be nice if the data/statistic chosen result in a BCa interval that is substantially different from the `'perc'` and `'basic'` versions. Also, if you could provide the bootstrap replicates of the statistic from R (`boot.out$t`), I could check the BCa interval that I'd get with the same bootstrap distribution. There might also be other functions in the `boot` package that could provide relevant output to compare against.

#### Additional information
We would have supported multi-sample statistics in the original PR, but we were unsure how the jackknife should be performed (to approximate the acceleration parameter) when there are multiple samples.

~~In this PR, the jackknife removes one observation from _one_ sample at a time because this seems to be the recommendation in [Missing Data, Imputation, and the Bootstrap](https://www.jstor.org/stable/2290846?seq=1). This is in contrast to what would seem like the more careful (and _much_ more time-consuming) option: removing one observation from _every_ sample (in each possible way), which is the recommendation of [Quantifying effects in two-sample environmental experiments using bootstrap confidence intervals](https://www.sciencedirect.com/science/article/abs/pii/S1364815205002239?casa_token=OcSrOn8nclMAAAAA:FKgCGSdCysMAjQl6Qf7v58encmDQmdSgOdV3iv1uFI48ePUgbz7HjpzaYBKMF9bRs7v4Ll0OGcc). However, that article mentions performing a comparison with the fast method implemented here, and they found "only very small differences (vs. nested sums) in case of robust measures and small differences in case of non-robust measures, not affecting the main conclusions of this paper". I'll continue to look into the theory, but @steppi if you know or are interested in the theory of the bootstrap, this is (yet) another one I could use some help with!~~

Efron presents the calculation of the acceleration parameter in the multi-sample case in Chapter 15.4 of An Introduction to the Bootstrap, specifically Equation 15.36.

See https://github.com/scipy/scipy/issues/16433#issuecomment-1161358545 for additional discussion.

### CODE REVIEW

File: scipy/stats/_resampling.py [hunk 1]
```diff
@@ -103,17 +103,27 @@ def _percentile_along_axis(theta_hat_b, alpha):
...
+    for i, sample in enumerate(data):
+        # jackknife resampling will add an axis prior to the last axis that
+        # corresponds with the different jackknife resamples. Do the same for
+        # each sample of the data to ensure broadcastability. We need to
+        # create a copy of the list containing the samples anyway, so do this
+        # in the loop to simplify the code. This is not the bottleneck...
+        samples = [np.expand_dims(sample, -2) for sample in data]
+        for jackknife_sample in _jackknife_resample(sample, batch):
+            samples[i] = jackknife_sample
+            samples = _broadcast_arrays(samples, axis=-1)
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:125]: `samples` needs a different name in here.

**Comment by FlorinAndrei** [comment]
I'm dealing with a time crunch right now, I'm not sure if I can help. I'll add a comment here if that changes. Sorry!

File: scipy/stats/_resampling.py [hunk 2]
```diff
@@ -103,17 +103,27 @@ def _percentile_along_axis(theta_hat_b, alpha):
...
+            theta_hat_i.append(statistic(*samples, axis=-1))
     theta_hat_i = np.concatenate(theta_hat_i, axis=-1)
     theta_hat_dot = theta_hat_i.mean(axis=-1, keepdims=True)
     num = ((theta_hat_dot - theta_hat_i)**3).sum(axis=-1)
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:129]: These are not quite right. The individual terms from different samples should not be weighted equally.

**Comment by steppi** [comment]
> I'll continue to look into the theory, but @steppi if you know or are interested in the theory of the bootstrap, this is (yet) another one I could use some help with!

Looks interesting. I’ll take a look over the weekend.

**Comment by mdhaber** [comment]
@steppi I think I corrected this based on equation 15.36 of An Introduction to the Bootstrap. The code attempts to mirror the math in the paper, so hopefully it's not too hard to follow. 

One thing that I considered using as a unit test is Efron's example calculation of the acceleration parameter from the text. Unfortunately, I can't get my numbers to match his : ) I'm not certain that the mistake is mine, though. I'm pretty confident because:

* The acceleration parameter I calculate for one-sample statistics using this PR ([line 143](https://github.com/scipy/scipy/blob/e05695fb045b832f09c234f6d035587c39dd8f00/scipy/stats/_resampling.py#L143)) is the same as the acceleration parameter I get in `main` ([line 122](https://github.com/scipy/scipy/blob/b894251498625ba5140743394884c44281b87274/scipy/stats/_resampling.py#L122)).
* A very straightforward implementation of Efron's example (included below) gives the same value of the acceleration parameter as this PR.

So I think either I'm misinterpreting the math in the equation, I'm mixing up the data or statistic used in the example, or the numerical result in the book is not quite right.

If you could help check my work - perhaps by implementing your own calculation of the equation 15.36 - I'd really appreciate it!

<details>

Here is the equation in question:
image

At the end, he gives a value of $\hat{a} = 0.06$ for some statistic $\hat{\theta}_0$ and some data. 

A few pages earlier, the statistic $\hat{\theta}_0$ appears to be defined as the difference between two sample means:
image

And I think the example was performed on the "mouse data" from Table 15.1 / Table 2.1:
image

A very straightforward implementation of the calculation is here:
```python3
import numpy as np

y = [10, 27, 31, 40, 46, 50, 52, 104, 146]
z = [16, 23, 38, 94, 99, 141, 197]

def statistic(z, y, axis=0):
    return np.mean(z, axis=axis) - np.mean(y, axis=axis)

n = len(z)
m = len(y)

theta_z_i = []
for i in range(n):
    zi = z[:i] + z[i+1:]
    theta_z_i.append(statistic(zi, y))
    
theta_y_i = []
for i in range(m):
    yi = y[:i] + y[i+1:]
    theta_y_i.append(statistic(z, yi))
    
theta_z_i = np.array(theta_z_i)
theta_y_i = np.array(theta_y_i)

theta_z_dot = np.sum(theta_z_i) / n
theta_y_dot = np.sum(theta_y_i) / m

U_z_i = (n - 1) * (theta_z_dot - theta_z_i)
U_y_i = (m - 1) * (theta_y_dot - theta_y_i)

num = np.sum(U_z_i**3) / n**3 + np.sum(U_y_i**3) / m**3
den = np.sum(U_z_i**2) / n**2 + np.sum(U_y_i**2) / m**2

a_hat = 1/6 * num/den**(3/2)  # 0.011008228344026734
```
 
If we execute:
```python3
from scipy import stats
data = (z, y)
stats.bootstrap(data, statistic)
```
in debug mode and inspect the value of `a_hat` (which is not returned by the function), it has the same value.

</details>

**Comment by steppi** [comment]
> If you could help check my work - perhaps by implementing your own calculation of the equation 15.36 - I'd really appreciate it!

Sure! I should have some time this coming weekend.

**Comment by mdhaber** [comment]
OK! I have a test that does not clearly show this is wrong!

When bootstrapping the difference between skewnesses of two skewnorm distributions, SciPy and wBoot BCa methods agree. Furthermore, there is a clear distinction between the three different methods (so the agreement isn't because, say, BCa is the same as the other methods and the other methods agree with wBoot).

Here's the SciPy code:
<details>
import numpy as np
from scipy import stats

m, n = 7, 8
method = 'basic'

rng = np.random.default_rng(0)
skew1 = stats.skewnorm(1)
skew2 = stats.skewnorm(2)

x = skew1.rvs(size=m, random_state=rng)
y = skew2.rvs(size=n, random_state=rng)

def statistic(x, y, axis):
    s1 = stats.skew(x, axis=axis)
    s2 = stats.skew(y, axis=axis)
    return s1 - s2
    
ref = skew1.stats('s') - skew2.stats('s')

res = stats.bootstrap((x, y), statistic, method=method, axis=-1, batch=100)
a, b = res.confidence_interval
print(a, b, res.standard_error)
# print(np.mean([a, b]))
</details>

Here is the CI low, CI high, and standard error for each of the three methods:
```
Percentile: -3.04659090989246    0.5429236319778981   0.9238995381294645
BCa:        -3.3341025959302417  0.1775610397990395   0.9260825841011494
Basic:      -3.5861151007055416 -0.059122444701895205 0.9039647046062937
```
(For the CIs, of course there is some variability between runs, but they are always within 0.1 of these values.)

And for the R `wBoot` package `boot.two.per` and `boot.two.bca`:
```
Percentile:  -3.041              0.4816               0.9093398
BCa:         -3.281              0.1772               0.9088609 
```

Maybe the best way to sum it up is with the midpoint of the confidence interval.
(SciPy) percentile: -1.2518, BCa: -1.5782, percentile: -1.8226
(wBoot) percentile: -1.2797, BCa: -1.5519

R code can be run at https://rdrr.io/snippets/
<details>
```
library(wBoot)
library(moments)

x = c( 0.75859206,  0.5910282 , -0.4419409 , -0.36654601,  0.34955357, -1.38835871,  0.76735821)
y = c(1.41186073, 0.49775975, 0.08275588, 0.24086388, 0.03567057,
       0.52024419, 0.31966611, 1.32067634)

twoskew<- function(x1, y1) {
skewness(x1) - skewness(y1)
}
boot.two.bca(x, y, skewness, conf.level = 0.95, R = 9999, stacked = FALSE)
```
</details>

**Comment by steppi** [comment]
@mdhaber, hey I haven't forgotten about this, just in a very busy period now. I hope to look into it over the weekend. Please ping me on anything else that needs a look. 

My schedule should stabilize in October, and I should have a free couple hours per week to contribute afterwards. Things are tough now though.

**Comment by mdhaber** [comment]
No problem. Thanks!

**Review by steppi — APPROVED** [review]
OK, I'm fairly confident your implementation is correct. I double checked the mouse data example in Section 15.4 of _An Introduction to to the Bootstrap_ and got the same result as you. I also read through Efron's 1994 paper _Missing Data, Imputation, and the Bootstrap_, and think the mythical remark J can be found at the end of Section 5 pasted below
image

Applying this general idea, (but using  bootstrap resamples and jackknife estimates instead of the numerical derivatives used in that section of the paper), yields the formula you've used, which agrees with that of 15.36 in the bootstrap textbook when the number of samples is 2.

I'm not sure where the discrepancy between our result and Efron's come from, perhaps it was just a numerical error.

This looks good to me, so approving.

**Comment by mdhaber** [comment]
Thanks so much for looking into this! I'll add the mouse data acceleration parameter example and the comparison against R as a test, then I'll be done with this. After that, would you like to merge?

**Comment by steppi** [comment]
> Thanks so much for looking into this! I'll add the mouse data acceleration parameter example and the comparison against R as a test, then I'll be done with this. After that, would you like to merge?

No problem. Sounds good!

**Comment by mdhaber** [comment]
@steppi I added the tests. I think CI failures are just because I didn't merge main (and CircleCI doesn't do that automatically). LMK if you'd like me to do that and re-run CI before merging. Otherwise, I think this is good to go.

**Comment by steppi** [comment]
@mdhaber Oh sorry. Just getting to this now. I saw that the only docs you touched built correctly so figured it was OK to merge as is.
